includes/AMP/Traits/Sanitization_Utils.php

```diff
@@ -0,0 +1,182 @@
+			}
+			if ( ! $hyperlink->getAttribute( 'rel' ) ) {
+				$hyperlink->setAttribute( 'rel', 'noreferrer' );
+			}
+		}
+	}
+
+	/**
+	 * Replaces the amp-story end tag to include amp-analytics tag if set up.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @param Document|AMP_Document $document Document instance.
+	 * @return void
+	 */
+	private function insert_analytics_configuration( &$document ) {
+		/* @var DOMElement $story_element The <amp-story> element. */
+		$story_element = $document->body->getElementsByTagName( 'amp-story' )->item( 0 );
+
+		if ( ! $story_element ) {
+			return;
+		}
+
+		ob_start();
+
+		/**
+		 * Fires before the closing <amp-story> tag.
+		 *
+		 * Can be used to print <amp-analytics> configuration.
+		 *
+		 * @since 1.1.0
+		 */
+		do_action( 'web_stories_print_analytics' );
+
+		$output = (string) ob_get_clean();
+
+		if ( empty( $output ) ) {
+			return;
+		}
+
+		$fragment = $document->createDocumentFragment();
+		$fragment->appendXml( $output );
+
+		$story_element->appendChild( $fragment );
+	}
+
+	/**
+	 * Replaces the placeholder of publisher logo in the content.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @param Document|AMP_Document $document       Document instance.
+	 * @param string                $publisher_logo Publisher logo.
+	 * @param string                $placeholder    Placeholder publisher logo.
+	 * @return void
+	 */
+	private function add_publisher_logo( &$document, $publisher_logo, $placeholder ) {
+		/* @var DOMElement $story_element The <amp-story> element. */
+		$story_element = $document->body->getElementsByTagName( 'amp-story' )->item( 0 );
+
+		if ( ! $story_element ) {
+			return;
+		}
+
+		// Add a publisher logo if missing or just a placeholder.
+		$existing_publisher_logo = $story_element->getAttribute( 'publisher-logo-src' );
+
+		if ( ! $existing_publisher_logo || $existing_publisher_logo === $placeholder ) {
+			$story_element->setAttribute( 'publisher-logo-src', $publisher_logo );
+		}
+
+		// Without a publisher logo, a story becomes invalid AMP.
+		// Remove the 'amp' attribute to not mark it as an AMP document anymore,
+		// preventing errors from showing up in GSC and other tools.
+		if ( ! $story_element->getAttribute( 'publisher-logo-src' ) ) {
+			$document->html->removeAttribute( 'amp' );
+		}
+	}
+
+	/**
+	 * Adds square, and landscape poster images to the <amp-story>.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @param Document|AMP_Document $document      Document instance.
+	 * @param string[]              $poster_images List of poster images, keyed by type.
+	 * @return void
+	 */
+	private function add_poster_images( &$document, $poster_images ) {
+		/* @var DOMElement $story_element The <amp-story> element. */
+		$story_element = $document->body->getElementsByTagName( 'amp-story' )->item( 0 );
+
+		if ( ! $story_element ) {
+			return;
+		}
+
+		foreach ( $poster_images as $attr => $url ) {
+			$story_element->setAttribute( $attr, esc_url( $url ) );
```

Comment:
Should use `esc_url_raw()` because no HTML escaping should be performed in a DOM context.

```suggestion
			$story_element->setAttribute( $attr, esc_url_raw( $url ) );
```